servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/DefaultGrpcClientCallFactory.java

```diff
@@ -204,14 +231,33 @@ public Completable onClose() {
         return streamingHttpClient.onClose();
     }
 
-    private static <Req> HttpRequest newAggregatedRequest(final GrpcClientMetadata metadata, final Req rawReq,
-                                                          final HttpRequestFactory requestFactory,
-                                                          final GrpcSerializationProvider serializationProvider,
-                                                          final List<ContentCodec> supportedCodings,
-                                                          final Class<Req> requestClass) {
+    private <Req> HttpRequest newAggregatedRequest(final GrpcClientMetadata metadata, final Req rawReq,
+                                                   final HttpRequestFactory requestFactory,
+                                                   final GrpcSerializationProvider serializationProvider,
+                                                   final List<ContentCodec> supportedCodings,
+                                                   final Duration timeout, final Class<Req> requestClass) {
         final HttpRequest httpRequest = requestFactory.post(metadata.path());
-        initRequest(httpRequest, supportedCodings);
+        initRequest(httpRequest, supportedCodings, timeout);
         return httpRequest.payloadBody(uncheckedCast(rawReq),
                 serializationProvider.serializerFor(metadata.requestEncoding(), requestClass));
     }
+
+    /**
+     * Determines the timeout for a new request using three potential sources; the client default, the client metadata
+     * or the async context. The lowest timeout will be used.
+     *
+     * @param metaDataTimeout the timeout specified in client metadata
+     * @return The timeout {@link Duration}, potentially negative.
+     */
+    private Duration timeoutForRequest(Duration metaDataTimeout) {
```

Comment:
Hm, I think the manually provided `metaDataTimeout` should override the default value configured on the builder, when `GRPC_DEADLINE_KEY` is empty. Users may have higher timeout expectation for a specific endpoint, while others are fine with lower default value.

The `GRPC_DEADLINE_KEY` should have the highest priority if it's the lowest value, because otherwise the service filter will trigger cancellation.

`defaultTimeout` is more like a fallback if non of the above is set. It's not "maxTimeout", it's just a default.

WDYT about this order of priorities: `min(Duration.between(Instant.now(), AsyncContext.get(GRPC_DEADLINE_KEY)), metaDataTimeout) || defaultTimeout`?

Reply:
Because if there is a `GRPC_DEADLINE_KEY ` value in async context it means that we are inside a service which has a timeout filter and that timeout will be triggered regardless of what user explicitly set on the meta-data. Therefore, there is not reason to propagate the higher value that user provided to downstream services if the current service is going to timeout earlier.

Reply:
I am fine with the suggestion.